src/engine/Match_tainting_mode.ml

```diff
@@ -132,6 +132,84 @@ type propagator_match = {
 }
 (** Taint will flow from `from` to `to_` through the axiliary variable `id`. *)
 
+(* This Formula_tbl structure is used to create a "formula cache", which will
+   permit the sharing of matches resulting from common formulas that occur as
+   sources, sinks, sanitizers, or propagators.
+
+   In particular, because of hardcoded propagators, we expect to see lots of
+   sharing from Semgrep Pro Engine.
+*)
+module Formula_tbl = struct
+  include Hashtbl.Make (struct
+    type t = Rule.formula
+
+    let equal = AST_utils.with_structural_equal Rule.equal_formula
+    let hash = Rule.hash_formula
+  end)
+
+  let cached_find_opt formula_cache formula compute_matches_fn =
+    match find_opt formula_cache formula with
+    | None ->
+        (* if it's not in the formula table at all, then don't
+           bother caching, just compute and move on
+
+           See [get_shared_formulas] above.
+        *)
+        compute_matches_fn ()
+    | Some None ->
+        (* if it's in the table, but is not yet cached, then
+           compute and add it into the table.
+        *)
+        let ranges, expls = compute_matches_fn () in
+        replace formula_cache formula (Some (ranges, expls));
+        (ranges, expls)
+    | Some (Some (ranges, expls)) -> (ranges, expls)
+end
+
+(* This function is for creating a formula cache which only caches formula that
+   it knows will be shared, at least once, among the formula in a bunch of
+   taint rules.
+
+   This is because it's obviously not useful to cache a formula's matches if
+   that formula never comes up again. This cache stores an option, with keys
+   that are only formula that are guaranteed to appear more than once in the
+   collection.
+*)
+let mk_specialized_formula_cache (rules : Rule.taint_rule list) =
+  let count_tbl = Formula_tbl.create 128 in
+  let flat_formulas =
+    rules
+    |> List.concat_map (fun rule ->
+           let (`Taint (spec : R.taint_spec)) = rule.R.mode in
+           Common.map (fun source -> source.R.source_formula) (snd spec.sources)
+           @ Common.map
+               (fun sanitizer -> sanitizer.R.sanitizer_formula)
+               spec.sanitizers
+           @ Common.map (fun sink -> sink.R.sink_formula) (snd spec.sinks)
+           @ Common.map
+               (fun propagator -> propagator.R.propagator_formula)
+               spec.propagators)
+  in
+  flat_formulas
+  |> List.iter (fun formula ->
+         match Formula_tbl.find_opt count_tbl formula with
+         | None -> Formula_tbl.add count_tbl formula 1
+         | Some x -> Formula_tbl.replace count_tbl formula (1 + x));
+  (* This new table will map each formula which exhibits sharing
```

Comment:
I would go further and keep the actual number of occurrences. Then, whenever the formula is requested via `cached_find_opt` I would decrease this counter. And when the counter reaches `0` then we know it can be freed.

The simpler alternative, is that we require a formula to occur in X% of the rules before we share it. If a formula is just used in two rules, and we have 200 rules to run, and the formula produces a lot of findings (hence consumes a lot of memory), then it may not be worthwhile to cache it. We could experiment with different percentages and find some sweet spot.

Reply:
Here's a summary of the sharing counts in `pro` with `p/default:
```
!!!This is a proprietary extension of semgrep.!!!
!!!You should not call directly this program.!!!
here's a table
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
2
2
2
2
2
19
19
19
19
19
19
19
19
19
19
19
19
19
19
19
19
19
19
19
19
19
19
19
19
here's a table
1
1
1
here's a table
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
3
here's a table
1
1
1
1
1
1
1
1
1
1
1
1
2
2
4
here's a table
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
2
2
2
2
2
19
19
19
19
19
19
19
19
19
19
19
19
19
19
19
19
19
19
19
19
19
19
19
19
here's a table
1
1
1
1
2
3
3
3
3
3
3
3
3
3
here's a table
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
2
2
2
2
2
2
2
2
2
2
2
2
2
2
2
2
4
4
4
4
5
7
19
20
62
62
62
62
62
62
62
62
62
here's a table
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
2
3
here's a table
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
2
2
2
2
2
2
3
3
3
6
17
here's a table
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
2
3
5
6
8
8
9
here's a table
1
1
1
1
3
here's a table
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
here's a table
1
1
1
2
2
2
2
2
2
2
2
2
2
2
2
2
2
2
2
2
```
where each "here's a table" denotes a new invocation of `mk_specialized_formula_cache`. I used it on a single target, so I'm not actually quite sure why it's being called so many times, but it should be being called for each batch of taint rules, per language. So maybe that's why.

This seems to show some measure of sharing. Notably `p/default` has about 23 `interfile: true` rules, so those should be the ones with hardcoded propagators, I think.